> 문제1) 로지스틱 회귀분석

In [29]:
import pandas as pd 
import numpy as np 
from scipy.stats import pearsonr
from sklearn.linear_model import LinearRegression, LogisticRegression

In [30]:
import seaborn as sns
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [31]:
# 분석 데이터 결정 
# 독립변수 : sex, sibsp , fare
# 종속변수 
df = df[['survived','sex','sibsp','fare']]
df.head()

survived     sex  sibsp     fare
0         0    male      1   7.2500
1         1  female      1  71.2833
2         1  female      0   7.9250
3         1  female      1  53.1000
4         0    male      0   8.0500

In [32]:
# 변수가 문자형인 경우 변환 해야한다
# from sklearn.preprocessing import LabelEncoder
# le = LabelEncoder()
# df['sex'] = le.fit_transform(df['sex'])

In [33]:
df['sex'] = df['sex'].map({'female':1,'male':0})
df

survived  sex  sibsp     fare
0           0    0      1   7.2500
1           1    1      1  71.2833
2           1    1      0   7.9250
3           1    1      1  53.1000
4           0    0      0   8.0500
..        ...  ...    ...      ...
886         0    0      0  13.0000
887         1    1      0  30.0000
888         0    1      1  23.4500
889         1    0      0  30.0000
890         0    0      0   7.7500

[891 rows x 4 columns]

In [34]:
# LogisticeRegression() 객체안에는 반드시 penalty='none' 옵션 
model = LogisticRegression(penalty='none')
model.fit(df.iloc[:,1:],df.iloc[:,0])

LogisticRegression(penalty='none')

In [35]:
print(model.coef_)
print(model.intercept_)

[[ 2.56677766 -0.40170461  0.01375665]]
[-1.69644546]


> 문제2) 로지스틱 회귀모형에서 sibsp 변수가 한 단위 증가할 때 생존할 오즈가 몇배 증가하는지 구하시오
- 반올림 하여 소수점 셋째자리까지 구하시오

__결과 : Logit = ln(P/1-P) = -1.6964 + 2.5668sex - 0.4017sibp + 0.0138fare__

In [36]:
odds_ratio = round(np.exp(model.coef_[0,1]),3)
print(odds_ratio)

0.669


> 문제3) 로지스틱 회귀모형에서 여성일 경우 남성에 비해 오즈가 몇 배 증가하는지 구하시오     
(1 값을 어떤 것으로 잡았냐에 따라 달라진다!!!)

In [39]:
# 여자가 1일 경우 
odds_ratio=round(np.exp(model.coef_[0,0]),3)
print(odds_ratio)

13.024


In [41]:
import statsmodels.api as sm

x=df.iloc[:,1:]
y=df.iloc[:,0]
x=sm.add_constant(x)

model=sm.Logit(y,x).fit()
summary = model.summary()
print(summary)

Optimization terminated successfully.
         Current function value: 0.483846
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      887
Method:                           MLE   Df Model:                            3
Date:                Thu, 28 Nov 2024   Pseudo R-squ.:                  0.2734
Time:                        21:30:01   Log-Likelihood:                -431.11
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 5.094e-70
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.6964      0.129    -13.134      0.000      -1.950      -1.443
sex            2.5668      0.